### PR TITLE
Rework syntax highlighting

### LINKED ISSUE #4: Inline C-style comments aren't highlighted
image

### LINKED ISSUE #5: Hex numbers aren't highlighted
image

### LINKED ISSUE #7: Comments are not highlighted in the first line of a struct typedef
image

image

image

image

### LINKED ISSUE #8: The first element of a struct typedef is highlighted with a different color
image

### LINKED ISSUE #9: Members are not highlighted in a multi-line struct
image

### LINKED ISSUE #10: Keywords in function names confuse the highlighting
image

### LINKED ISSUE #11: Keywords in comments confuse the highlighting
image

### PR DESCRIPTION
**Pull request by antroseco**
Fixes #4 
image

Fixes #5 
image

Fixes #7, fixes #8, fixes #9
image

Fixes #10 
image

Fixes #11 
image

### CODE REVIEW

File: syntaxes/graphene.tmLanguage.json [hunk 1]
```diff
@@ -25,8 +25,8 @@
 			"name": "constant.language.graphene"
 		},
 		{
-			"comment": "Floats",
-			"match": "\\b\\d+(\\.\\d+)?([eE][+-]?\\d+)?\\b",
+			"comment": "Decimal constants",
+			"match": "\\b\\d+(\\.\\d+)?\\b",
```
Comment (DaveDuck321) [inline syntaxes/graphene.tmLanguage.json]: Does this mean we're not going to support exponential notation?

File: syntaxes/graphene.tmLanguage.json [hunk 2]
```diff
@@ -179,14 +162,44 @@
...
+				},
+				{
+					"comment": "String literals",
+					"name": "string.quoted.double.graphene",
+					"begin": "\"",
+					"end": "\"",
+					"patterns": [
+						{
+							"name": "constant.character.escape.graphene",
+							"match": "\\\\."
```
Comment (DaveDuck321) [inline syntaxes/graphene.tmLanguage.json:111]: That's super readable regex is fun...

File: syntaxes/graphene.tmLanguage.json [hunk 3]
```diff
@@ -89,47 +89,49 @@
...
-		{
-			"comment": "Return type in function declarations",
-			"begin": "(?<=(function|foreign|@operator|@implicit).*?(-\\>))",
-			"end": "=|$",
-			"patterns": [
+					"begin": "\\(",
+					"end": "->",
+					"patterns": [
+						{
+							"include": "#comments"
```
Comment (DaveDuck321) [inline syntaxes/graphene.tmLanguage.json]: Oh damn, do you just need to include this everywhere that comments can go?

File: syntaxes/graphene.tmLanguage.json [hunk 4]
```diff
@@ -206,14 +208,32 @@
...
+						},
+						{
+							"match": "\\b\\w+\\b",
+							"name": "variable.parameter.graphene"
+						}
+					]
+				},
+				{
+					"begin": "(?<=:)",
+					"end": ",|\\)|}",
```
Comment (DaveDuck321) [inline syntaxes/graphene.tmLanguage.json]: Do you also need to match against `]` for array indexing... Or is this only used in top-level constructs?

File: syntaxes/graphene.tmLanguage.json [hunk 5]
```diff
@@ -254,6 +210,63 @@
...
+							"include": "#comments"
+						},
+						{
+							"include": "#argument-list"
+						}
+					]
+				},
+				"function-return-type": {
+					"begin": "(?<=->)",
+					"end": "(?<=\\w)(?=\\s)",
```
Comment (DaveDuck321) [inline syntaxes/graphene.tmLanguage.json]: Does is still work if the return type is adhoc?
Let's say `function a : () -> {a : int} = {...}`

**Review by DaveDuck321 — COMMENTED** [review]
This looks much more maintainable than before. Can't say I understand all the regex but at least it's modular now.

On thing I noticed is that in the screenshot example the return type isnt highlighted, not sure if you've fix it since then tho.

Looking forward to using this next time I'm working on the parser 😊.

File: syntaxes/graphene.tmLanguage.json (hunk 1 above)
Comment (antroseco) [reply syntaxes/graphene.tmLanguage.json]: No reason not to, I'll just revert this change when we do end up implementing it

File: syntaxes/graphene.tmLanguage.json (hunk 2 above)
Comment (antroseco) [reply syntaxes/graphene.tmLanguage.json:111]: Yeah... json makes it 2x worse unfortunately

File: syntaxes/graphene.tmLanguage.json (hunk 3 above)
Comment (antroseco) [reply syntaxes/graphene.tmLanguage.json]: Basically, yes.

TextMate syntax highlighting only operates on single lines, and once a rule matches, it consumes those characters. To get around the single-line limitation, and parse stuff like function declarations spread over multiple lines, you need to use `begin` and `end` rules, which match the `begin` regex and then keep looking at other lines until they match `end`. Once you match both `begin` and `end`, you stop looking at the other top-level rules (like comments) for that section of text.

The way to process that section of text further is with nested `patterns`. These rules are run again in the matched span of text, so we can match for stuff like comments and argument lists.

File: syntaxes/graphene.tmLanguage.json (hunk 4 above)
Comment (antroseco) [reply syntaxes/graphene.tmLanguage.json]: The `}` is there for stuff like `function foo : (arg: {value: int}) -> void`. Not sure if we should even highlight array indexing.

File: syntaxes/graphene.tmLanguage.json (hunk 5 above)
Comment (antroseco) [reply syntaxes/graphene.tmLanguage.json]: Nope, forgot that was a thing!

**Comment by antroseco** [comment]
> On thing I noticed is that in the screenshot example the return type isnt highlighted, not sure if you've fix it since then tho.

Ah nice spot. That seems to break whenever you have adhoc structs in the arguments.

EDIT: fixed in 713cbdb1a6c522fc33a98840c40966a6d7c69f7e

File: syntaxes/graphene.tmLanguage.json (hunk 5 above)
Comment (antroseco) [reply syntaxes/graphene.tmLanguage.json]: This works now :)

**Comment by antroseco** [comment]
I'm gonna merge now to do a release, but feel free to add more comments

**Pull request merged**